### PR TITLE
Pt3t3rsupport

### PR DESCRIPTION
**Pull request by biswajitSM**
@tritemio Thanks for going through the previous PR and the suggestions. Clearly I am a novice here. And sorry for unnecessarily taking your time.
Here I have tried to implement the changes in a readable way.
Please feel free to ignore the PR if it is a "rubbish" again. I really needed the support for pt3 and t3r saving. If it is easy for you, you may also add the feature instead of converging the changes from the fork.

I have really benefited from many of your repositories.
Thanks a lot!
Biswajit

### CODE REVIEW

**Comment by biswajitSM** [comment]
I think it is having problem finding the data.

**Comment by tritemio** [comment]
@biswajitSM, thanks, now the PR looks better. I'll go over it the next few days and let you know.

**Comment by biswajitSM** [comment]
@tritemio thanks a lot. I forgot to add overflow correction function for .t3r functions:

def process_t3records_t3rfile(t3records.....
    ....
    _t3r_correct_overflow(timestamps, valid, overflow_ch, overflow)
    return detectors, timestamps, nanotimes
def _t3r_correct_overflow(timestamps, valid, overflow_ch, overflow):
    """Apply overflow correction when each overflow has a special timestamp.
    """
    overflow_correction = 0;
    for i in xrange(valid.size):
        if valid[i] == 0:
            overflow_correction += overflow
        timestamps[i] += overflow_correction

**Comment by tritemio** [comment]
@biswajitSM, just add another commit and push it to github, it will appear here. Thanks.

File: phconvert/pqreader.py [hunk 1]
```diff
@@ -643,13 +822,29 @@ def _correct_overflow_nsync(timestamps, detectors, overflow_ch, overflow):
...
         timestamps[i] += overflow_correction
-
+def _t3r_correct_overflow(timestamps, valid, overflow_ch, overflow):
+    """Apply overflow correction when each overflow has a special timestamp.
+    """
+    overflow_correction = 0;
+    for i in xrange(valid.size):
+        if valid[i] == 0:
+            overflow_correction += overflow
+        timestamps[i] += overflow_correction
```
Comment (tritemio) [inline phconvert/pqreader.py]: This function appears identical to `_correct_overflow1` with detectors renamed to valid. If this is the case please remove and use  `_correct_overflow1`  instead.

File: phconvert/pqreader.py [hunk 2]
```diff
@@ -643,13 +822,29 @@ def _correct_overflow_nsync(timestamps, detectors, overflow_ch, overflow):
...
+def _t3r_correct_overflow2(timestamps, valid, overflow_ch, overflow):
+    """Apply overflow correction when each overflow has a special timestamp.
+    """
+    index_overflows = np.where((valid == 0))
+    num_overflows = timestamps[index_overflows]
+    cum_overflows = np.zeros(timestamps.size, dtype='int64')
+    cum_overflows[index_overflows] = num_overflows
+    np.cumsum(cum_overflows, out=cum_overflows)
+    timestamps += (cum_overflows * overflow)
```
Comment (tritemio) [inline phconvert/pqreader.py]: This function is very similar to `_correct_overflow_nsync`. The argument `detectors` is renamed `valid` and the argument `overflow_ch` is not used anymore. I suggest just using the previous function. Passing `overflow_ch = 0` should achieve the same effect.

File: phconvert/pqreader.py [hunk 3]
```diff
@@ -607,6 +752,40 @@ def process_t3records(t3records, time_bit=10, dtime_bit=15,
...
+    if special_bit:
+        ch_bit += 1
+    assert ch_bit <= 8
+    assert time_bit <= 16
+    assert time_bit+reserved+valid+dtime_bit+ch_bit == 32
+    
+    detectors = np.bitwise_and(
+        np.right_shift(t3records, time_bit+dtime_bit+reserved+valid), 2**ch_bit-1).astype('uint8')
+    nanotimes = np.bitwise_and(
+    np.right_shift(t3records, dtime_bit), 2**time_bit-1).astype('uint16')
```
Comment (tritemio) [inline phconvert/pqreader.py]: please add indentation on this line (continuation lines require indentation)

File: phconvert/pqreader.py [hunk 4]
```diff
@@ -607,6 +752,40 @@ def process_t3records(t3records, time_bit=10, dtime_bit=15,
...
+    dt = np.dtype([('low16', 'uint16'), ('high16', 'uint16')])
+    t3records_low16 = np.frombuffer(t3records, dt)['low16']     # View
+    timestamps = t3records_low16.astype(np.int64)               # Copy
+    np.bitwise_and(timestamps, 2**dtime_bit - 1, out=timestamps)
+
+    overflow_ch = 2**ch_bit - 1
+    overflow = 2**dtime_bit
+    #_t3r_correct_overflow = numba.jit('void(i8[:], u1[:], u4, u8)')(
+    #    _t3r_correct_overflow)
+    _t3r_correct_overflow(timestamps, valid, overflow_ch, overflow)
```
Comment (tritemio) [inline phconvert/pqreader.py]: please remove commented lines

**Comment by tritemio** [comment]
@biswajitSM, thanks again for your contribution. I finally had time to reviewed your patch. It is in quite good, but I'm requiring a few modifications and clarifications before I can merge it. See the detailed comments above.

Also, can you provide the example data files in t3r and pt3 format that you used in the notebook. We need to add them to the pool of data files used by CI for testing. In this way the tests on AppVeyor and Travis will have the chance to pass. To upload the files you can drag-and-drop them in the comment box here on github.

File: phconvert/loader.py [hunk 5]
```diff
@@ -268,6 +268,170 @@ def nsalex_ht3(filename,
...
+    creation_time = time.strftime("%Y-%m-%d %H:%M:%S", ctime_t)
+
+    provenance = dict(
+        filename = filename,
+        creation_time = creation_time,
+        software = software,
+        software_version = software_version,
+    )
+
+
```
Comment (tritemio) [inline phconvert/loader.py]: remove empty line

File: phconvert/loader.py [hunk 6]
```diff
@@ -268,6 +268,170 @@ def nsalex_ht3(filename,
...
+        excitation_wavelengths = excitation_wavelengths,
+        excitation_cw = [False, False],
+        detection_wavelengths = detection_wavelengths,
+        excitation_alternated = [False, False])
+
+    data = dict(
+        _filename=filename,
+        acquisition_duration = acquisition_duration,
+
+        photon_data = photon_data,
```
Comment (tritemio) [inline phconvert/loader.py]: remove empty line

File: phconvert/loader.py [hunk 7]
```diff
@@ -268,6 +268,170 @@ def nsalex_ht3(filename,
...
+    creation_time = time.strftime("%Y-%m-%d %H:%M:%S", ctime_t)
+
+    provenance = dict(
+        filename = filename,
+        creation_time = creation_time,
+        software = software,
+        software_version = software_version,
+    )
+
+
```
Comment (tritemio) [inline phconvert/loader.py]: remove one empty line

File: phconvert/loader.py [hunk 8]
```diff
@@ -268,6 +268,170 @@ def nsalex_ht3(filename,
...
+        lifetime = True,
+        excitation_wavelengths = excitation_wavelengths,
+        excitation_cw = [False, False],
+        detection_wavelengths = detection_wavelengths,
+        excitation_alternated = [False, False])
+
+    data = dict(
+        _filename=filename,
+        acquisition_duration = acquisition_duration,
+
```
Comment (tritemio) [inline phconvert/loader.py]: remove empyty line

File: phconvert/pqreader.py [hunk 9]
```diff
@@ -455,6 +480,126 @@ def ptu_reader(filename):
...
+    with open(filename, 'rb') as f:
+        # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+        # Binary file header
+        header_dtype = np.dtype([
+                ('Ident',             'S16'   ),
+                ('SoftwareVersion',     'S6'    ),
+                ('HardwareVersion',     'S6'    ),
+                #('CreatorName',       'S18'   ),
+                #('CreatorVersion',    'S12'   ),
+                ('FileTime',          'S18'   ),
```
[the remark this answers is not in the document]
Comment (biswajitSM) [reply phconvert/pqreader.py]: done!

File: phconvert/pqreader.py [hunk 10]
```diff
@@ -607,6 +752,40 @@ def process_t3records(t3records, time_bit=10, dtime_bit=15,
         ovcfunc = _correct_overflow
     ovcfunc(timestamps, detectors, overflow_ch, overflow)
     return detectors, timestamps, nanotimes
+def process_t3records_t3rfile(t3records, reserved=1, valid=1, time_bit=12, dtime_bit=16,
+                      ch_bit=2, special_bit=False):
```
[the remark this answers is not in the document]
Comment (biswajitSM) [reply phconvert/pqreader.py]: done!

File: phconvert/pqreader.py (hunk 3 above)
Comment (biswajitSM) [reply phconvert/pqreader.py]: done!

File: phconvert/pqreader.py (hunk 4 above)
Comment (biswajitSM) [reply phconvert/pqreader.py]: done!

File: phconvert/pqreader.py (hunk 1 above)
Comment (biswajitSM) [reply phconvert/pqreader.py]: Indeed. It works with _correct_overflow1 by setting the overflow_ch=0. The extra functions is removed.

File: phconvert/pqreader.py (hunk 2 above)
Comment (biswajitSM) [reply phconvert/pqreader.py]: The function is removed now.

File: phconvert/loader.py (hunk 5 above)
Comment (biswajitSM) [reply phconvert/loader.py]: done!

File: phconvert/loader.py (hunk 6 above)
Comment (biswajitSM) [reply phconvert/loader.py]: done!

File: phconvert/loader.py (hunk 7 above)
Comment (biswajitSM) [reply phconvert/loader.py]: done!

File: phconvert/loader.py (hunk 8 above)
Comment (biswajitSM) [reply phconvert/loader.py]: done!

**Comment by biswajitSM** [comment]
Thanks a lot again for going through in detail. This is turning out to be a steep learning process.
I have acted on all your comments. Please let me know if something is undone.
It doesn't support the file extensions. So here they are zipped and attached:
[data.zip](https://github.com/Photon-HDF5/phconvert/files/1336330/data.zip)

**Comment by tritemio** [comment]
@biswajitSM your doing great for a first contribution👍 This is the normal code-review process. I'm just making sure that the code is clear and easy to maintain once is merged. We are converging, not quite done yet though, bear with me. 😉 

Now that you uploaded the data I was able to run the notebooks and I have a few more questions.

1. The T3R notebook gives an error:

```
~/src/phconvert/phconvert/pqreader.py in t3r_reader(filename)
    570         # array elements are actually present in the file is indicated by
    571         # ImgHdrSize above.
--> 572         ImgHdr = np.fromfile(f, dtype='int32', count=ttmode['ImgHdrSize'])
    573 
    574         # The remainings are all T3 records

TypeError: only integer scalar arrays can be converted to a scalar index
```
New versions of numpy are more stringent. Now `count` is required to be a scalar but it is a 1-element array. To fix add a `[0]` (e.g. `count=ttmode['ImgHdrSize']` with `count=ttmode['ImgHdrSize'][0]`) both for ImgHdrSize and nRecords (next line). I fixed this bug a while ago in the HT3 and PTU functions, but your code was based on an earlier version so you didn't get the fix.

2. Looking at t3r_reader I hadn't noticed that you set timestamps_unit and nanotimes_unit manually, not from metadata? Is this general? Do you have a link to the t3r documentation so we can figure out the right way to do it?

3. The PT3 notebook runs without errors. However it contains only one detector. Is it because it is a test file? If you have the ability to produce a PIE (also called nsALEX) data file that would be less confusing for a new user running the notebook. Otherwise what's the point of calling the notebook nsALEX and having a load_nsalex function?

**Comment by biswajitSM** [comment]
hi @tritemio , Sorry for the extremely late reply.
1. the t3r reading didn't show probel for me though it use to give warning. But now with a tiny modification in 't3fr_reader', it works fine.
2. I couldn't find a way to extract the timestamp_unit and nanotimes_unit from the meta data. here is a documentation about t3r. Data format is explained on page-58.
[Manual_timeharp200.pdf](https://github.com/Photon-HDF5/phconvert/files/1380333/Manual_timeharp200.pdf)

3. I use phconvert mostly to read picoquant file and then analyze in different ways. But we have a a setup for alex_FRET and my colleagues will be doing FRET soon. Here is a .pt3 data file with two detectors and I suppose with alternating excitation with 488 nm and 635 nm.

[DNA_FRET_0.5nM.pt3.zip](https://github.com/Photon-HDF5/phconvert/files/1380341/DNA_FRET_0.5nM.pt3.zip)

Thanks again for all the review.
Biswajit

**Comment by tritemio** [comment]
@biswajitSM sorry for the late reply. Thanks for the modifications and new data files. I fixed a few typos and added CI integration so that your notebooks are tested. I'm only waiting for CI tests to pass before merging.

**Comment by tritemio** [comment]
@biswajitSM, I forgot, we need of course to add your name to the contributor list 😃 

https://github.com/Photon-HDF5/phconvert#authors--contributors

Let me now which name/email you want to list there.

**Comment by tritemio** [comment]
@biswajitSM, another question regarding the data files. I saw you uploaded the first two to figshare. Would you mind uploading also the last one. I would rather download from figshare than from github for the CI tests.

**Comment by tritemio** [comment]
All test passed, merging.

**Comment by biswajitSM** [comment]
Thanks for the merge. I have added the last data file to figshare.
Though I am more of a beneficiary of the repo, still it feels good to be a part of open source. :D Here is my name and email address:
Biswajit Pradhan
biswajitp145@gmail.com

**Comment by tritemio** [comment]
@biswajitSM, I added you to the list of authors. Thanks again for your contribution, looking forward to the next one 😉

**Pull request merged**